Source/WebCore/rendering/style/TextUnderlineOffset.h

```diff
@@ -26,66 +26,78 @@
 #pragma once
 
 #include "Length.h"
+#include <variant>
 #include <wtf/text/TextStream.h>
 
 namespace WebCore {
 
 class TextUnderlineOffset {
 public:
-    static constexpr TextUnderlineOffset createWithAuto()
+    struct Auto {
+        bool operator==(const Auto&) const = default;
+    };
+
+    static TextUnderlineOffset createWithAuto()
     {
-        return { std::nullopt };
+        return TextUnderlineOffset(Auto { });
     }
 
-    static constexpr TextUnderlineOffset createWithLength(float length)
+    static TextUnderlineOffset createWithLength(Length&& length)
     {
-        return { length };
+        return TextUnderlineOffset(WTFMove(length));
     }
 
-    constexpr bool isAuto() const
+    bool isAuto() const
     {
-        return !isLength();
+        return std::holds_alternative<Auto>(m_value);
     }
 
-    constexpr bool isLength() const
+    bool isLength() const
     {
-        return static_cast<bool>(m_length);
+        return std::holds_alternative<Length>(m_value);
     }
 
-    void setLengthValue(float length)
+    const Length& length() const
     {
         ASSERT(isLength());
-        m_length = length;
+        return std::get<Length>(m_value);
     }
 
-    constexpr float lengthValue() const
+    float resolve(float fontSize, float defaultValue = 0.f) const
     {
+        if (isAuto())
+            return defaultValue;
+
         ASSERT(isLength());
-        return *m_length;
+        const Length& len = std::get<Length>(m_value);
+        if (len.isPercent())
+            return fontSize * (len.percent() / 100.0f);
+        if (len.isCalculated())
+            return len.nonNanCalculatedValue(fontSize);
+        return len.value();
     }
 
-    constexpr float lengthOr(float defaultValue) const
+    bool operator==(const TextUnderlineOffset& other) const = default;
+private:
+    std::variant<Auto, Length> m_value;
```

Comment:
Length has a type for Auto (`Length(LengthType::Auto)`). I don't think the variant is needed.